Source/monster.cpp

```diff
@@ -1294,28 +1297,23 @@ void MonsterAttackMonster(int i, int mid, int hper, int mind, int maxd)
 	auto &monster = Monsters[mid];
 	assert(monster.MType != nullptr);
 
-	if (monster._mhitpoints >> 6 > 0 && (monster.MType->mtype != MT_ILLWEAV || monster._mgoal != MGOAL_RETREAT)) {
+	if (!monster.IsPossibleToHit()) {
 		int hit = GenerateRnd(100);
 		if (monster._mmode == MonsterMode::Petrified)
 			hit = 0;
-		bool unused;
-		if (!CheckMonsterHit(monster, &unused) && hit < hper) {
+		if (monster.TryLiftGargoyle())
+			return;
+		if (hit < hper) {
 			int dam = (mind + GenerateRnd(maxd - mind + 1)) << 6;
 			monster._mhitpoints -= dam;
 			if (monster._mhitpoints >> 6 <= 0) {
-				if (monster._mmode == MonsterMode::Petrified) {
-					StartDeathFromMonster(i, mid);
+				StartDeathFromMonster(i, mid);
+				if (monster._mmode == MonsterMode::Petrified)
 					monster.Petrify();
-				} else {
-					StartDeathFromMonster(i, mid);
-				}
 			} else {
-				if (monster._mmode == MonsterMode::Petrified) {
-					MonsterHitMonster(mid, i, dam);
+				MonsterHitMonster(mid, i, dam);
+				if (monster._mmode == MonsterMode::Petrified)

```

Comment:
`MonsterHitMonster()` won't change `_mmode`, no need for this call.

Reply:
As above. but I am not sure what id does with `AnimInfo.IsPetrified` part of Petrify(); Don't want to go to deep to monster in this PR.

Reply:
I looked in to it as part of the reviews, if you just make the remaining suggested changes everything should be good and clean. No more calls to `monster.Petrify()` :)

Reply:
Yes, `_mmode` isn't changed. But here I am not sure if that Petrify() sets `AnimInfo.IsPetrified` isn't needed here just in case.
I would like to not to mess with it right now.

Reply:
NewMonsterAnim is the one that changes IsPetrified, so since `MonsterHitMonster` never calls that for petrified monsters there is no need for it here. This is actually the one place where the check was always unnecessary :)

Reply:
As previous conv above ^ Finally I was convinced to revemo it as you suggested. (commit soon)

Reply:
> NewMonsterAnim is the one that changes IsPetrified, so since `MonsterHitMonster` never calls that for petrified monsters there is no need for it here. This is actually the one place where the check was always unnecessary :)

Yes, I am probably too much phobic about it. I am afraid there may be some strange case when `_mode` is Petrified, but anim info` AnimInfo.IsPetrified` was not set yet to true.